File: Client/Assets/Common/Scripts/Net/Protocol/ArrayProxy.cs
#region Header
/**
 * 名称：NumberProxy
 
 * 日期：2015.11.16
 * 描述：
 **/
#endregion
using UnityEngine;
using System;
using System.Collections;
using System.Collections.Generic;

namespace NetCore
{
    public class ArrayProxy : Proxy
    {
        #region Fields

        #endregion


        #region Properties

        #endregion


        #region Constructors


        public ArrayProxy()
        {

        }
        #endregion


        #region Static Methods

        #endregion


        #region Private Methods

        #endregion

        public override bool getValue(IoBuffer ioBuffer, System.Type type, byte flag, out object value)
        {
            value = null;
            if (!type.IsArray)
            {
                Debuger.LogError("不是数组类型");
                return false;
            }

            int len;
            if (!readVarInt32(ioBuffer, out len))
            {
                return false;
            }

            Type elemType = type.GetElementType();
            Array arr = Array.CreateInstance(elemType, len);

            BeginParentLog("数组", len);
            bool needNewLine = elemType.IsClass && typeof(string) != elemType.GetElementType();

            for (int i = 0; i < len; i++)
            {
                if (needNewLine) AddLogNewLine("");
                object e;
                if (!ProtocolCoder.instance.Decode(ioBuffer, elemType, out e))
                    return false;
                arr.SetValue(e, i);
                AddLog(",");//加个分隔符
            }
            value = arr;
            EndParentLog();

            return true;
        }

        public override bool setValue(IoBuffer ioBuffer, object value)
        {
            byte flag = Types.ARRAY;

            // #### 0000
            ioBuffer.Write(flag);
            System.Array arr = (System.Array)value;
            if (arr == null)
            {
                Debuger.LogError("不能转为array类型：{0}", value.GetType().Name);
                return false;
            }

            BeginParentLog("数组", arr.Length);
            bool needNewLine = value.GetType().GetElementType().IsClass && typeof(string) != value.GetType().GetElementType();
            putVarInt32(ioBuffer, arr.Length);
            for (int i = 0; i < arr.Length; i++)
            {
                if (needNewLine) AddLogNewLine("");
                ProtocolCoder.instance.Encode(ioBuffer, arr.GetValue(i));
                AddLog(",");//加个分隔符
            }
            EndParentLog();

            return true;
        }
    }
}

File: Server/GameServer/app/libs/protocolProxy/arrayProxy.js
"use strict";

var BaseProxy = require("./baseProxy").BaseProxy;
var ProtocolCoder = require("../protocolCoder");

class ArrayProxy extends BaseProxy
{
    /**
     * 读取数据
     * @param {IOBuffer} ioBuf - 数据缓冲
     * @param {number} flag - 类型标记
     * @return {Array} 读取的数据
     * @override
     */
    getValue(ioBuf, flag) {
        //var typeFlag = BaseProxy.getFlagType(flag);
        //if (typeFlag !== BaseProxy.TYPE_FLAG_ARRAY) {
        //    throw new Error("ArrayProxy~getValue 必须是Array类型");
        //}

        var coder = ProtocolCoder.instance;

        //这里可能会抛出错误，让上层处理
        var len = BaseProxy.readVarUInt(ioBuf);

        var arr = new Array(len);

        for (var i = 0; i < len; ++i) {
            arr[i] = coder.decode(ioBuf);
        }

        return arr;
    }

    /**
     * 写入数据
     * @param {IOBuffer} ioBuf - 数据缓冲
     * @param {Array} value - 数据
     * @override
     */
    setValue(ioBuf, value)
    {
        //if (!Object.isArray(value))
        //{
        //    throw new Error("ArrayProxy~setValue 必须是Array类型");
        //}

        var coder = ProtocolCoder.instance;

        ioBuf.writeUInt8(BaseProxy.TYPE_FLAG_ARRAY);
        var len = value.length;
        BaseProxy.writeVarUInt(ioBuf, len);
        for (var i = 0; i < len; ++i) {
            coder.encode(ioBuf, value[i]);
        }
    }
}

exports.ArrayProxy = ArrayProxy;